### PR TITLE
Deprecate calendar ops

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes most of the #1010 issue.
- [x] Tests for the changes have been added (for bug fixes / features) (mostly removed)
  - [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGES.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Calendar utilities that have an equivalent in xarray have been deprecated and will be removed in 0.51. This concerns the following members of ``xclim.core.calendar``:
    - ``convert_calendar`` : Use ``Dataset.convert_calendar``, ``DataArray.convert_calendar`` or ``xr.coding.calendar_ops.convert_calendar``  instead.
        + If your code passes ``target`` as an array, first convert the source to the target's calendar and then reindex the result to ``target``.
        + If you were using the ``doy=True`` option, replace it with ``xc.core.calendar.convert_doy(source, target_cal).convert_calendar(target_cal)``.
        + ``"default"`` is no longer a valid calendar name for any xclim functions and will not be returned by ``get_calendar``. Xarray has a ``use_cftime`` argument, xclim exposes it when the distinction is needed.
    - ``date_range`` : Use ``xarray.date_range`` instead.
    - ``date_range_like``: Use ``xarray.date_range_like`` instead.
    - ``interp_calendar`` : Use ``xarray.coding.calendar_ops.interp_calendar`` instead.
    - ``days_in_year`` : Use ``xarray.coding.calendar_ops._days_in_year`` instead.
    - ``datetime_to_decimal_year`` : Use ``xarray.coding.calendar_ops._datetime_to_decimal_year`` instead.

### Does this PR introduce a breaking change?

Yes.

- Some options of  `convert_calendar` are not possible with xarray, suggestions are made above.
- "default" is not a calendar name anymore.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -6,6 +6,19 @@ v0.50.0 (unreleased)
 --------------------
 Contributors to this version: Trevor James Smith (:user:`Zeitsperre`).
 
+Breaking changes
+^^^^^^^^^^^^^^^^
+* Calendar utilities that have an equivalent in xarray have been deprecated and will be removed in 0.51. (:issue:`1010`, :pull:`1761`). This concerns the following members of ``xclim.core.calendar``:
```
Comment (Zeitsperre) [inline CHANGES.rst:11]:
```suggestion
* Calendar utilities that have an equivalent in `xarray` have been deprecated and will be removed in `xclim` v0.51.0. (:issue:`1010`, :pull:`1761`). This concerns the following members of ``xclim.core.calendar``:
```

File: CHANGES.rst [hunk 2]
```diff
@@ -6,6 +6,19 @@ v0.50.0 (unreleased)
...
+    - ``convert_calendar`` : Use ``Dataset.convert_calendar``, ``DataArray.convert_calendar`` or ``xr.coding.calendar_ops.convert_calendar``  instead.
+        + If your code passes ``target`` as an array, first convert the source to the target's calendar and then reindex the result to ``target``.
+        + If you were using the ``doy=True`` option, replace it with ``xc.core.calendar.convert_doy(source, target_cal).convert_calendar(target_cal)``.
+        + ``"default"`` is no longer a valid calendar name for any xclim functions and will not be returned by ``get_calendar``. Xarray has a ``use_cftime`` argument, xclim exposes it when the distinction is needed.
+    - ``date_range`` : Use ``xarray.date_range`` instead.
+    - ``date_range_like``: Use ``xarray.date_range_like`` instead.
+    - ``interp_calendar`` : Use ``xarray.coding.calendar_ops.interp_calendar`` instead.
+    - ``days_in_year`` : Use ``xarray.coding.calendar_ops._days_in_year`` instead.
+    - ``datetime_to_decimal_year`` : Use ``xarray.coding.calendar_ops._datetime_to_decimal_year`` instead.
```
Comment (Zeitsperre) [inline CHANGES.rst:20]: Do we want to keep this one around, seeing that it's not formally exposed in xarray yet?

File: xclim/core/calendar.py [hunk 3]
```diff
@@ -83,13 +81,30 @@
...
+def _get_usecf_and_warn(calendar: str, xcfunc: str, xrfunc: str):
+    if calendar == "default":
+        calendar = "standard"
+        use_cftime = False
+        msg = " and use use_cftime=False instead of calendar='default' to get numpy objects."
+    else:
+        use_cftime = None
+        msg = ""
+    warn(
+        f"xclim function {xcfunc} is deprecated in favor of {xrfunc} and will be removed in 0.51. Please adjust your script{msg}.",
```
Comment (Zeitsperre) [inline xclim/core/calendar.py:93]: ;)

```suggestion
        f"`xclim` function {xcfunc} is deprecated in favour of {xrfunc} and will be removed in v0.51.0. Please adjust your script{msg}.",
```

File: xclim/core/calendar.py [hunk 4]
```diff
@@ -169,21 +148,18 @@ def get_calendar(obj: Any, dim: str = "time") -> str:
     Returns
     -------
     str
-      The cftime calendar name or "default" when the data is using numpy's or python's datetime types.
+      The cf calendar name.
```
Comment (Zeitsperre) [inline xclim/core/calendar.py:151]:
```suggestion
      The Climate and Forecasting (CF) calendar name.
```

File: xclim/core/calendar.py [hunk 5]
```diff
@@ -337,235 +330,44 @@ def convert_calendar(
...
-        out = out.where(out[dim].notnull(), drop=True)
-
-    if isinstance(target, str) and missing is not None:
-        target = date_range_like(source[dim], cal_tgt)
-
     if isinstance(target, xr.DataArray):
-        out = out.reindex(
-            {dim: target}, fill_value=missing if missing is not None else np.nan
+        raise NotImplementedError(
+            "In xclim 0.50, convert_calendar is only a copy of xarray.coding.calendar_ops.convert_calendar. "
```
Comment (Zeitsperre) [inline xclim/core/calendar.py:340]:
```suggestion
            "In `xclim` v0.50.0, `convert_calendar` is a direct copy of `xarray.coding.calendar_ops.convert_calendar`. "
```

File: xclim/core/calendar.py [hunk 6]
```diff
@@ -337,235 +330,44 @@ def convert_calendar(
...
         )
-
-    # Copy attrs but change remove `calendar` is still present.
-    out[dim].attrs.update(source[dim].attrs)
-    out[dim].attrs.pop("calendar", None)
-
-    return out
+    if doy is not False:
+        raise NotImplementedError(
+            "In xclim 0.50, convert_calendar is only a copy of xarray.coding.calendar_ops.convert_calendar. "
```
Comment (Zeitsperre) [inline xclim/core/calendar.py:345]:
```suggestion
            "In `xclim` v0.50.0, `convert_calendar` is a direct copy of `xarray.coding.calendar_ops.convert_calendar`. "
```

**Review by Zeitsperre — APPROVED** [review]
Looks good!

File: CHANGES.rst [hunk 7]
```diff
@@ -9,6 +9,16 @@ Contributors to this version: Trevor James Smith (:user:`Zeitsperre`).
...
+* Calendar utilities that have an equivalent in xarray have been deprecated and will be removed in 0.51. (:issue:`1010`, :pull:`1761`). This concerns the following members of ``xclim.core.calendar``:
+    - ``convert_calendar`` : Use ``Dataset.convert_calendar``, ``DataArray.convert_calendar`` or ``xr.coding.calendar_ops.convert_calendar``  instead.
+        + If your code passes ``target`` as an array, first convert the source to the target's calendar and then reindex the result to ``target``.
+        + If you were using the ``doy=True`` option, replace it with ``xc.core.calendar.convert_doy(source, target_cal).convert_calendar(target_cal)``.
+        + ``"default"`` is no longer a valid calendar name for any xclim functions and will not be returned by ``get_calendar``. Xarray has a ``use_cftime`` argument, xclim exposes it when the distinction is needed.
+    - ``date_range`` : Use ``xarray.date_range`` instead.
+    - ``date_range_like``: Use ``xarray.date_range_like`` instead.
+    - ``interp_calendar`` : Use ``xarray.coding.calendar_ops.interp_calendar`` instead.
```
Comment (aulemahal) [inline CHANGES.rst:19]:
```suggestion
    - ``interp_calendar`` : Use ``Dataset.interp_calendar`` or ``xarray.coding.calendar_ops.interp_calendar`` instead.
```

File: CHANGES.rst (hunk 2 above)
Comment (aulemahal) [reply CHANGES.rst:21]: Yeah, I had this hesitation... I don't think xarray will expose it.

Xarray does not respect the usual python conventions for private/public objects. I don't think anything under `coding` is public API. But at the same time, the corresponding xclim function were also not publicised much. Even though we assumed them as a public API I would say we treated them with the same level of attention as xarray does for its (private) copies.

Long story short, I personally prefer removing this code.

File: xclim/core/calendar.py (hunk 6 above)
Comment (aulemahal) [inline xclim/core/calendar.py]:
```suggestion
            "In xclim 0.51, convert_calendar is only a copy of xarray.coding.calendar_ops.convert_calendar. "
```

File: CHANGES.rst (hunk 2 above)
Comment (Zeitsperre) [reply CHANGES.rst:21]: I trust your judgment here. Feel free to remove it. The new version will be released after our meeting tomorrow.
